### PR TITLE
WFLY-6981 Upgrade Weld to 2.4.0.Final

### PR DESCRIPTION
**Pull request by mkouba**
- also remove the dependency on com.google.guava:guava
- also remove unused WeldAnnotationDiscovery (http://docs.jboss.org/weld/reference/latest/en-US/html/ri-spi.html#annotationDiscovery)

https://issues.jboss.org/browse/WFLY-6981

### CODE REVIEW

**Comment by mkouba** [comment]
The windows build fails with strange compilation error:
> [org.wildfly:wildfly-weld] [ERROR] /C:/BuildAgent/work/a31d203e70e89f90/weld/src/main/java/org/jboss/as/weld/discovery/WeldAnnotationDiscovery.java:[36,31] package com.google.common.cache does not exist

Note that `WeldAnnotationDiscovery` class is removed in this pull request.

**Comment by mkouba** [comment]
@stuartwdouglas Could you please review this PR? Thanks!

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/main/module.xml [hunk 1]
```diff
@@ -37,8 +37,6 @@
     </resources>
 
     <dependencies>
-        <module name="ch.qos.cal10n" />
```
Comment (stuartwdouglas) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/main/module.xml]: I think this module and org.slf4j.ext are no longer used after this change?

**Comment by stuartwdouglas** [comment]
Looks good to me, other than the module comment above. If these modules are unused they should be removed.

**Comment by mkouba** [comment]
@stuartwdouglas Thanks! Is there a way to detect unused modules?

**Comment by stuartwdouglas** [comment]
No, I just happened to know that those two were only used by weld.

**Comment by bstansberry** [comment]
@mkouba This is what I use when checking that kind of thing:

```
taozi:wildfly bstansberry$ git grep ch.qos.cal10n | grep module.xml
feature-pack/src/main/resources/modules/system/layers/base/ch/qos/cal10n/main/module.xml:<module xmlns="urn:jboss:module:1.3" name="ch.qos.cal10n">
feature-pack/src/main/resources/modules/system/layers/base/ch/qos/cal10n/main/module.xml:        <artifact name="${ch.qos.cal10n:cal10n-api}"/>
feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/main/module.xml:        <module name="ch.qos.cal10n" />
feature-pack/src/main/resources/modules/system/layers/base/org/jboss/weld/core/main/module.xml:        <module name="ch.qos.cal10n" />
feature-pack/src/main/resources/modules/system/layers/base/org/jboss/weld/probe/main/module.xml:        <module name="ch.qos.cal10n" />
feature-pack/src/main/resources/modules/system/layers/base/org/slf4j/ext/main/module.xml:        <module name="ch.qos.cal10n" />
```

**Comment by bstansberry** [comment]
However, ch.qos.cal10n is used by org.slf4j.ext and that module does not have the "private" jboss.api property in module.xml so removing it is a breaking change.

**Comment by mkouba** [comment]
@bstansberry So it can't be removed, right?

**Comment by bstansberry** [comment]
@mkouba Correct; it should stay. If @jamezp wants to remove it he should work out how to properly do that; e.g. perhaps add jboss.api="deprecated" for a while.  Also. we're keeping the EAP 7.x branch and WF master in fairly close sync at the moment so for now at least any deprecation/removal plan should be based on upgrading to the next minor (EAP 7.1) as opposed to a major (WildFly 11).

**Comment by stuartwdouglas** [comment]
AFAIK org.slf4j.ext was only added for weld. It is not added to the deployment as a dependency by any DUP's, so I think it not being marked private may have been an oversight.

**Comment by jamezp** [comment]
The `org.slf4j.ext` module has been available since AS 7. I think it was likely not marked as private because `org.slf4j` was not and the versions should be aligned in most cases. However  `org.slf4j.ext` is not added to deployments. I don't know if users are adding dependencies to it so I think we would have to deprecate it first.

There is one place [in core](https://github.com/wildfly/wildfly-core/blob/3.0.0.Alpha7/logging/src/main/java/org/jboss/as/logging/LoggingExtension.java#L100) where the module is used, but we can remove that when we remove the module itself.

**Comment by mkouba** [comment]
@stuartwdouglas @jamezp @bstansberry In any case, removing the modules is a different story - this PR is about Weld upgrade and subsystem cleanup. We're going to release 2.4.0.Final in a few days, so I can update this PR and rename the original issue (WFLY-6981).

**Comment by mkouba** [comment]
Closing in favor of https://github.com/wildfly/wildfly/pull/9185

**Pull request closed without merging**